### PR TITLE
Sme scoring for cn5

### PR DESCRIPTION
**Pull request by luminoso-beaudoin**
First, this PR adds an (optional) validation set to the training data; if present validation error will be computed and printed every 5000 training iterations.

Second, there are new functions for using a trained model to produce scores for ConceptNet edges.  This includes two new functions (score_edges and make_edge_score_file) and restoring evaluate_conceptnet to full working order (and changing its input from an EdgeDataset back to a file).

Third, I cleaned up a few miscellaneous issues (e.g. eliminating no-longer-needed module imports).

### CODE REVIEW

**Comment by luminoso-beaudoin** [comment]
[SME validation sets and edge scoring.](https://trello.com/c/pt6gODTy/571-sme-validation-sets-and-edge-scoring)

**Comment by luminoso-beaudoin** [comment]
As of now (2018/09/06) this PR incorporates both the changes I've made for getting scores from the SME model into a form useable in ConceptNet builds, and Rob's ACE-based training.  There are also miscellaneous improvements (e.g. generating a log file during training), support for tracking loss on a validation dataset during training, and functions to compute other metrics on the trained model.

File: conceptnet_sme/source_scoring.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    """
+    with open(path, 'rb') as fp:
+        canonical_scores = msgpack.unpack(fp, use_list=False, encoding='utf-8')
+    scores = {source: (np.float32(s0), np.float32(s1))
+              for source, (s0, s1) in canonical_scores.items()}
+    return scores
+
+
+if __name__ == '__main__':
+    from conceptnet5.util import get_data_filename
```
Comment (rspeer) [inline conceptnet_sme/source_scoring.py]: This import should be at the top of the file.

File: conceptnet_sme/semantic_matching.py [hunk 2]
```diff
@@ -420,44 +478,225 @@ def forward(self, rels, terms_L, terms_R):
...
+        input_batch = torch.empty(
+            [3, batch_size], dtype=torch.int64, requires_grad=False)
+        input_batch.pin_memory() # speed up any transfer to gpu
+        input_batch_on_device = torch.empty_like(input_batch, device=device)
+        input_batch_edges = []
+
+        position_in_batch = 0
+        for edge in edges:
+            # Translate the string fields of the edge to indices the model knows.
+            rel = edge[0]
```
Comment (rspeer) [inline conceptnet_sme/semantic_matching.py]: Minor thing: this could be written more simply as `rel, left, right = edge[:3]`.

File: conceptnet_sme/semantic_matching.py [hunk 3]
```diff
@@ -513,7 +753,23 @@ def load_model(filename):
         total_accum.print("Total time to load model:")
         return model
 
-    def evaluate_conceptnet(self, dataset, cutoff_value=-1, output_filename=None):
+    def evaluate_edge(self, rel_idx, left_idx, right_idx):
```
Comment (rspeer) [inline conceptnet_sme/semantic_matching.py]: I'm guessing that this could be sped up a fair amount if it could evaluate a batch of edges instead of just one edge at a time, right?

File: conceptnet_sme/semantic_matching.py [hunk 4]
```diff
@@ -966,40 +1273,45 @@ def model_copier(model, device):
...
+            collate_fn=validation_dataset.collate_batch,
+            sampler=CyclingSampler(validation_dataset),
+            pin_memory=True
+        )
+        # Since we want to step through the (infinite, cyclic) sequence of
+        # validation batches generated by this data loader, and take fixed-
+        # size slices from that sequence, we convert the data loader (which
+        # is iterable) into an iterator.  Note that (as the sequence is
+        # infinite) this iterator will never raise StopIteration.
+        validation_data_iterator = iter(validation_data_loader)
```
Comment (rspeer) [inline conceptnet_sme/semantic_matching.py]: Is this being too clever with the way we iterate over the data? It seems to me that we'd want the validation set to be a fixed, finite set.

File: conceptnet_sme/source_scoring.py (hunk 1 above)
Comment (luminoso-beaudoin) [reply conceptnet_sme/source_scoring.py]: Fixed.

File: conceptnet_sme/semantic_matching.py (hunk 2 above)
Comment (luminoso-beaudoin) [reply conceptnet_sme/semantic_matching.py]: Fixed.

File: conceptnet_sme/semantic_matching.py (hunk 3 above)
Comment (luminoso-beaudoin) [reply conceptnet_sme/semantic_matching.py]: Based on earlier experience with make_edge_score_file, I thought the speed up would not be worth the extra code complexity.  But you motivated me to try it, and the speed gain is significant (when I tried before I think I only had one gpu, which might make a difference).  So now I've refactored both evaluate_conceptnet and evaluate_statistics to use score_edges (which works on batches).  That required some small changes to score_edges, and uncovered a bug in it (handling the last, short, batch could break on multiple gpu's) which I fixed.

File: conceptnet_sme/semantic_matching.py (hunk 4 above)
Comment (luminoso-beaudoin) [reply conceptnet_sme/semantic_matching.py]: I did it this way because not every edge from a validation file will necessarily survive the filters (e.g. the relation must be in COMMON_RELATIONS, both the terms must be in the embedding) we impose, so it's easiest to use a large validation file to be sure some edges are left.  But then using the whole validation set is slow.  If you don't find that convincing let me know, and I'll make the change you propose.

Comment (rspeer) [reply conceptnet_sme/semantic_matching.py]: I'd prefer to be more certain about what's in the validation set, even if we evaluate on a sub-sample of it at any given moment. For example, we could do the filtering first and save the result.

Comment (luminoso-beaudoin) [reply conceptnet_sme/semantic_matching.py]: OK, done.
